### ISSUE TITLE
:should-component-update isn't getting called.

### ISSUE THREAD

**Issue by aerohit** [reagent-project/reagent#107]
I have a component which I don't want to be re-rendered. So I have attached a function to :should-component-update which basically returns false all the times. But, the component gets re-rendered. To investigate the issue, I put in a print statement inside the attached function, but I never see the message. So I infer that it isn't getting called. The code is below, could you please point out if I am doing something wrong.

``` clojure
(defn summary-photo-component []
  (with-meta
    (fn [props]
      (let [url (image-url 370 (:profileimage @state/accommodation))]
        [:div.photo
         {:style {:background-image (str "url('" url "')")}}
         "Foto van de accomodatie"]))
    {:should-component-update (fn [this old-argv new-argv] (do (prn "Foo") false))}))
```

Thanks and regards,
Rohit

**Comment by marcloyal3** [comment]
I'm new with Reagent, so take it with a grain of salt...

I found that one way to prevent components from updating that don't need to be updated is to use fine grained cursors so each component has an isolated slice of global state and changes to the atom in one component won't cause update in all others that use the same Reagent atom. I think that as long as the global state has isolated slices that correspond to components (or component instances) e.g. using fine grained cursors or separate atoms under one namespace, then, ideally, we shouldn't worry about managing when Reagent is going to update React's virtual DOM. React only updates the real DOM when there is a difference between previous and current state of its virtual DOM.

**Comment by seancorfield** [comment]
@aerohit your top-level function and the enclosed function do not have the same arity -- perhaps that is your problem? How are you actually using `summary-photo-component`?

**Comment by mike-thompson-day8** [comment]
@aerohit:  This is not an answer to your question ... so your are aware ... that `with-meta` technique is old school.  There are neater ways these days (which might make clearer to you what the problem is):
https://github.com/Day8/re-frame/wiki/Creating-Reagent-Components#form-3-a-class-with-life-cycle-methods

**Comment by Frozenlock** [comment]
Programming: Where a few months old is *old school*.

**Comment by mike-thompson-day8** [comment]
@Frozenlock   :-)

**Comment by holmsand** [comment]
@aerohit You're returning a new component from another one, and that is currently only supported using `reagent.core/create-class` (mentioned in the link from @mike-thompson-day8). It should be, so that is definitely sort of a bug, but it will require some surgery of reagent to make it work…

**Comment by marcloyal3** [comment]
@holmsand I had it working at some point using the with-meta approach but then being a noob I have no idea what I did because I can't get it to work again...  Anyway, when I set the return value of should-component-update to true I noticed extra updates were occurring but again I don't remember the exact behavior. Also, I don't believe that it was passing in the previous and current state correctly since the equality test for components whose state hasn't changed was resulting in false... I think it would certainly be a fundamentally good thing to give user control over when the component updates since cursors may be "coarse" (as in providing access to part of the atom that is shared by other instances of the component, when using a component factory pattern and passing in options that specialize each instance, e.g. instance of input component for first name vs instance of input component for last name) and so when the end user changes the first name the last name filed will also update... In such instances, and other scenarios I'm sure, it would be good to allow each instance to decide if it should update or not. My solution was to pass "fine grained" cursors (that provide access to only first name or last name etc not the whole form data structure) and that solved the problem of unnecessary updates but made configuring the component a more terse activity (passing in fine grained cursors)
